komposable-architecture-test/src/main/java/com/toggl/komposable/test/TestStore.kt

```diff
@@ -0,0 +1,179 @@
+import com.toggl.komposable.architecture.ReduceResult
+import com.toggl.komposable.architecture.Reducer
+import com.toggl.komposable.architecture.Store
+import com.toggl.komposable.extensions.createStore
+import com.toggl.komposable.extensions.map
+import com.toggl.komposable.scope.DispatcherProvider
+import io.kotest.matchers.shouldBe
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.channels.Channel.Factory.BUFFERED
+import kotlinx.coroutines.flow.onCompletion
+import kotlinx.coroutines.flow.onStart
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.test.TestScope
+import kotlinx.coroutines.test.runCurrent
+import java.util.UUID
+
+class TestStore<State, Action : Any?>(
+    initialState: () -> State,
+    reducer: () -> Reducer<State, Action>,
+    dispatcherProvider: DispatcherProvider,
+    val testCoroutineScope: TestScope,
+) {
+    private val store: Store<State, TestReducer.TestAction<Action>>
+    val reducer: TestReducer<State, Action>
+    val state: State
+        get() = reducer.state
+
+    private val effectSubscriptionChannel = Channel<Unit>(capacity = BUFFERED)
+
+    init {
+        this.reducer = TestReducer(
+            reducer(),
+            initialState(),
+            testCoroutineScope,
+            effectSubscriptionChannel,
+        )
+        this.store = createStore(
+            initialState = this.state,
+            reducer = this.reducer,
+            dispatcherProvider = dispatcherProvider,
+            storeScopeProvider = { testCoroutineScope },
+        )
+    }
+
+    suspend fun send(action: Action, assert: ((state: State) -> State)? = null) {
+        if (reducer.receivedActions.isNotEmpty()) {
+            throw IllegalStateException("Cannot send actions after receiving actions")
+        }
+
+        val previousState = reducer.state
+        store.send(TestReducer.TestAction(TestReducer.TestAction.Origin.Send(action)))
+        effectSubscriptionChannel.receiveCatching()
+        testCoroutineScope.runCurrent()
+        val currentState = reducer.state
+
+        if (assert != null) {
+            currentState shouldBe assert(previousState)
+        }
+    }
+
+    suspend fun receive(
+        action: Action,
+        assert: ((state: State) -> State)? = null,
+    ) {
+        receive({ it == action }, assert)
+    }
+
+    suspend fun receive(
+        actionPredicate: (Action) -> Boolean,
+        assert: ((state: State) -> State)? = null,
+    ) {
+        if (reducer.inFlightEffects.isEmpty()) {
+            receiveAction(actionPredicate, assert)
+        } else {
+            testCoroutineScope.runCurrent()
+            receiveAction(actionPredicate, assert)
+            // wait for the in-flight effect to finish
+        }
+    }
+
+    private fun receiveAction(
+        matching: (Action) -> Boolean,
+        assert: ((state: State) -> State)?,
+    ) {
+        val action = reducer.receivedActions.firstOrNull { matching(it.first) }?.first
```

Comment:
should we also assert the number of actions received here? Finding only the first matching action could leave the last received action unasserted?

Reply:
Yeah, makes sense. This should be strict.